File: java/src/main/java/benchmarks/BrainfuckArray.java
package benchmarks;

import java.util.*;

public class BrainfuckArray extends Benchmark {

    static class Tape {
        private byte[] tape;
        private int pos = 0;

        Tape() {
            tape = new byte[30000];
        }

        byte get() {
            return tape[pos];
        }

        void inc() {
            tape[pos] = (byte)(tape[pos] + 1);
        }

        void dec() {
            tape[pos] = (byte)(tape[pos] - 1);
        }

        void advance() {
            pos++;
            if (pos >= tape.length) {

                byte[] newTape = new byte[tape.length + 1];
                System.arraycopy(tape, 0, newTape, 0, tape.length);
                tape = newTape;
            }
        }

        void devance() {
            if (pos > 0) {
                pos--;
            }
        }
    }

    static class Program {
        private final byte[] commands;
        private final int[] jumps;

        Program(String text) {

            int cmdCount = 0;
            for (int i = 0; i < text.length(); i++) {
                char c = text.charAt(i);
                if ("[]<>+-,.".indexOf(c) != -1) {
                    cmdCount++;
                }
            }

            commands = new byte[cmdCount];
            int idx = 0;
            for (int i = 0; i < text.length(); i++) {
                char c = text.charAt(i);
                if ("[]<>+-,.".indexOf(c) != -1) {
                    commands[idx++] = (byte)c;
                }
            }

            jumps = new int[commands.length];
            int[] stack = new int[commands.length];
            int sp = 0;

            for (int i = 0; i < commands.length; i++) {
                byte cmd = commands[i];
                if (cmd == '[') {
                    stack[sp++] = i;
                } else if (cmd == ']') {
                    if (sp > 0) {
                        int start = stack[--sp];
                        jumps[start] = i;
                        jumps[i] = start;
                    }
                }
            }
        }

        long run() {
            long result = 0L;
            Tape tape = new Tape();
            byte[] cmds = commands;
            int[] jmps = jumps;
            int pc = 0;

            while (pc < cmds.length) {
                byte cmd = cmds[pc];

                if (cmd == '[') {
                    if (tape.get() == 0) {
                        pc = jmps[pc];
                        continue;
                    }
                } else if (cmd == ']') {
                    if (tape.get() != 0) {
                        pc = jmps[pc];
                        continue;
                    }
                } else {
                    switch (cmd) {
                    case '+':
                        tape.inc();
                        break;
                    case '-':
                        tape.dec();
                        break;
                    case '>':
                        tape.advance();
                        break;
                    case '<':
                        tape.devance();
                        break;
                    case '.':
                        result = (result << 2) + (tape.get() & 0xFF);
                        break;
                    }
                }
                pc++;
            }

            return result;
        }
    }

    private String programText;
    private String warmupText;
    private long resultVal;

    public BrainfuckArray() {
        programText = Helper.configS(name(), "program");
        warmupText = Helper.configS(name(), "warmup_program");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "BrainfuckArray";
    }

    private long runProgram(String programText) {
        return new Program(programText).run();
    }

    @Override
    public void warmup() {
        long prepareIters = warmupIterations();
        for (long i = 0; i < prepareIters; i++) {
            runProgram(warmupText);
        }
    }

    @Override
    public void run(int iterationId) {
        resultVal += runProgram(programText);
    }

    @Override
    public long checksum() {
        return resultVal;
    }
}

File: scala/src/main/scala/benchmarks/BrainfuckArray.scala
package benchmarks

import scala.collection.mutable

class BrainfuckArray extends Benchmark:
  private val programText: String = Helper.configS(name(), "program")
  private val warmupText: String = Helper.configS(name(), "warmup_program")
  private var resultVal: Long = 0L

  override def name(): String = "BrainfuckArray"

  private def runProgram(programText: String): Long =
    BrainfuckArray.Program(programText).run()

  override def warmup(): Unit =
    val prepareIters = warmupIterations()
    for i <- 0L until prepareIters do
      runProgram(warmupText)

  override def run(iterationId: Int): Unit =
    resultVal += runProgram(programText)

  override def checksum(): Long = resultVal

object BrainfuckArray:
  class Tape:
    private var tape: Array[Byte] = new Array[Byte](30000)
    private var pos: Int = 0

    def get(): Byte = tape(pos)

    def inc(): Unit = tape(pos) = (tape(pos) + 1).toByte

    def dec(): Unit = tape(pos) = (tape(pos) - 1).toByte

    def advance(): Unit =
      pos += 1
      if pos >= tape.length then

        val newTape = new Array[Byte](tape.length + 1)
        Array.copy(tape, 0, newTape, 0, tape.length)
        tape = newTape

    def devance(): Unit =
      if pos > 0 then pos -= 1

  class Program(text: String):

    private val commands: Array[Byte] = {
      val chars = text.toCharArray
      val buffer = new Array[Byte](chars.length)  
      var count = 0

      for c <- chars do
        if "[]<>+-,.".indexOf(c) != -1 then
          buffer(count) = c.toByte
          count += 1

      val result = new Array[Byte](count)
      Array.copy(buffer, 0, result, 0, count)
      result
    }

    private val jumps: Array[Int] = {
      val arr = new Array[Int](commands.length)
      val stack = new Array[Int](commands.length)  
      var sp = 0  

      for i <- commands.indices do
        commands(i).toChar match
          case '[' =>
            stack(sp) = i
            sp += 1  
          case ']' =>
            if sp > 0 then
              sp -= 1
              val start = stack(sp)  
              arr(start) = i
              arr(i) = start
          case _ =>

      arr
    }

    def run(): Long =
      var result: Long = 0L
      val tape = Tape()
      var pc = 0
      val cmds = commands  
      val jmps = jumps

      while pc < cmds.length do
        cmds(pc).toChar match
          case '+' => tape.inc()
          case '-' => tape.dec()
          case '>' => tape.advance()
          case '<' => tape.devance()
          case '[' =>
            if tape.get() == 0 then pc = jmps(pc)
          case ']' =>
            if tape.get() != 0 then pc = jmps(pc)
          case '.' =>
            result = (result << 2) + (tape.get() & 0xFF)
          case _ =>
        pc += 1
      result